In [2]:
import pandas as pd

In [2]:
og = pd.read_csv('data/GoodReads_100k_books.csv')
#clean_book = pd.read_csv('data/after_encoding/clean_book_data.csv')
clean_book= pd.read_csv('data/translate_title.csv')

3 things to clean -> euro symbol in price, bookformat probably needs cleaning and genre (repeats)

In [17]:
def removeDuplicates(genre):
    genre_list = genre.split(',')
    print(genre_list)
    genre_list = list(dict.fromkeys(genre_list))
    if 'North American Hi...' in genre_list:
        genre_list.remove('North American Hi...')
        genre_list.append('North American History')
    elif 'Scandinavian Lite...' in genre_list:
        genre_list.remove('Scandinavian Lite...')
        genre_list.append('Scandinavian Literature')
    elif 'International Rel...' in genre_list:
        genre_list.remove('International Rel...')
        genre_list.append('International Relations')
    elif 'International Dev...' in genre_list:
        genre_list.remove('International Dev...')
        genre_list.append('International Development')
    elif 'Science Fiction R...' in genre_list:
        genre_list.remove('Science Fiction R...')
        genre_list.append('Science Fiction Romance')
    elif 'Democratic Republic Of The ...' in genre_list:
        genre_list.remove( 'Democratic Republic Of The ...')
        genre_list.append('Democratic Republic of the Congo')
    elif 'Complementary Med...' in genre_list:
        genre_list.remove('Complementary Med...')
        genre_list.append('Complementary Medicine')
    return ','.join(genre_list)

In [20]:
#clean_book['genre'] = clean_book['genre'].apply(removeDuplicates)
clean_book=pd.read_csv('data/no_extra_space_clean.csv')
clean_book['genre'][0]
#removeDuplicates(clean_book['genre'][0])
#'History,Military History,Civil War,American History,American Civil War,Nonfiction,Native Americans,North American History'


'History,Military History,Civil War,American History,American Civil War,Nonfiction,Native Americans,North American History'

In [6]:
#for i in clean_book.index:
 #   clean_book.loc[i,'price']=str(clean_book.loc[i,'price']).replace('�','€')

clean_book['price']

0        21,50 €
1        38,50 €
2            NaN
3            NaN
4            NaN
          ...   
72244        NaN
72245        NaN
72246        NaN
72247        NaN
72248    22,04 €
Name: price, Length: 72249, dtype: object

In [8]:
#clean_book['bookformat'] = clean_book['bookformat'].str.title()
clean_book['bookformat'].unique()

array(['Hardcover', 'Paperback', 'Audiobook', 'Mass Market Paperback',
       'Textbook', 'Board Books', 'Turtleback', 'Ebook',
       'Unknown Binding', 'Spiral-bound', 'Boxed Set', 'Other Format',
       'Trade Paperback', 'Novelty Book', 'Unbound', 'Library Binding',
       'Comics', 'Pamphlet', 'Spiral-Bound', 'Paper', 'Flexibound',
       'Game', 'Cards', 'Deck of 78 Tarot Cards', 'Plastic Comb',
       'Paperback with CD', 'Pocket', 'Leather Bound', 'Paperback Manga',
       'Illustrated Book and CD Set', 'Map', 'Misc. Supplies',
       'Light Novel', 'Diary', 'Tarot Deck & Book', 'Sheet music',
       'Flash Movie Webcast', 'Paperback And Deluxe Hardcover',
       'Ringbound', 'Paperback with Deck of Cards',
       'Paperback and Deluxe Hardcover', 'Webtoon'], dtype=object)

- Hardcover == Hardback == Pasta Dura == Capa Dura == Relié == Hardcover In Dust Jacket == Cloth Hardcover == Cloth == Hardcover, Slipcased == Over-Sized 12"W X 16"H Hardcover == Gebunden ==  'Hardcover ' == Gebundene Ausgabe == 単行本 == Inbunden == Casebound
- Paperback == Broschiert == Broché == Perfect Paperback == 文庫 == Capa Comum == Perfect == Sewn Paperback With Dust Jacket == Pasta Blanda == Saddle-Stitched Paperback == 120 X 190 Mm
- Paperback And Deluxe Hardcover == Hardcover, Paperback
- Mass Market Paperback == Mass_Market == Mass Market
- Textbook == Textbook Binding'
- Board Book == Board == Board Books
- Turtleback
- Ebook == Nook 
- Unknown Binding
- Audio Cd == Audio == Audio Cassette == Mp3 Book == Mp3 Cd == Audio Play == Audible Audio == Audiocd == Cd-Rom == Kindle Edition With Audio/Video == Audiobook
- Spiral-Bound == Spiral == Spiral=Bound
- Boxed Set == Boxed Set - Hardcover == Bookprint == Boxed Collection
- Other Format
- Trade Paperback
- Novelty Book
- Unbound
- Library Binding == Library
- Comic == Comics == コミック
- Dvd == Dvd (Ntsc)
- Pamphlet == Brochura
- Flexibound
- Game
- Paperback With Deck Of Cards
- Cards
- Deck Of 78 Tarot Cards == 78 Card Tarot Deck
- Plastic Comb
- Kindle Edition
- Leather Bound
- Paperback W/ Cd == Paperback With Cd Rom
- Taschenbuch (pocket book) == Poche == Softcover == Soft Cover == Pocket
- Paperback Manga
- Illustrated Book And Cd Set
- Map
- Misc. Supplies
- Light Novel
- Loose Leaf == Paper == Looseleaf
- Diary
- Tarot Deck & Book == Tarot Cards & Book == Tarot Deck And Booklet == Tarot Deck & Booklet
- Sheet Music
- Flash Movie Webcast
- 3-Ring Binder == Ringbound
- Webtoon

In [9]:
def map_bookformat(bookformat):
    if bookformat=='Hardback' or bookformat=='Pasta Dura' or bookformat=='Capa Dura' or bookformat=='Relié' or bookformat=='Hardcover In Dust Jacket' or bookformat=='Hardcover in dust jacket' or bookformat=='Cloth Hardcover' or bookformat=='Cloth' or bookformat=='cloth' or bookformat=='Hardcover, Slipcased' or bookformat=='Over-Sized 12"W X 16"H Hardcover' or bookformat=='Gebunden' or bookformat== 'Hardcover ' or bookformat=='Gebundene Ausgabe' or bookformat=='単行本' or bookformat=='Inbunden' or bookformat=='Casebound' or bookformat=='Slipcased Hardcover' or bookformat=='Pasta dura' or bookformat=='hardcover' or bookformat=='Capa dura' or bookformat=='Over-Sized 12"W x 16"H Hardcover' or bookformat=='hardback':
        return 'Hardcover'
    elif bookformat=='Broschiert' or  bookformat== 'Broché' or bookformat=='Perfect Paperback' or bookformat=='perfect' or bookformat=='文庫' or bookformat=='Capa Comum' or bookformat=='Perfect' or bookformat == 'Sewn Paperback With Dust Jacket' or bookformat == 'Pasta Blanda' or bookformat == 'Saddle-Stitched Paperback == 120 X 190 Mm' or bookformat=='paperback' or bookformat=='Capa comum' or bookformat=='Sewn paperback with dust jacket' or bookformat=='Pasta blanda' or bookformat=='Saddle-stitched paperback' or bookformat=='120 x 190 mm':
        return 'Paperback'
    elif bookformat=='Hardcover, Paperback':
        return 'Paperback And Deluxe Hardcover'
    elif bookformat=='Mass_Market' or bookformat=='Mass Market' or bookformat=='mass_market':
        return 'Mass Market Paperback'
    elif bookformat=='Textbook Binding':
        return 'Textbook'
    elif bookformat=='Board Book' or bookformat=='Board' or bookformat=='Board book':
        return 'Board Books'
    elif bookformat=='Nook' or bookformat=='Kindle Edition' or bookformat=='ebook':
        return 'Ebook'
    elif bookformat=='Audio Cd' or bookformat=='Audio' or bookformat=='Audio Cassette' or bookformat == 'Mp3 Book' or bookformat== 'Mp3 Cd' or bookformat == 'Audio Play' or bookformat == 'Audible Audio' or bookformat == 'Audiocd' or bookformat== 'Cd-Rom' or bookformat == 'Kindle Edition With Audio/Video' or bookformat=='Dvd' or bookformat== 'Dvd (Ntsc)' or bookformat=='Audio CD' or bookformat=='DVD (NTSC)' or bookformat=='MP3 Book' or bookformat=='MP3 CD' or bookformat=='DVD' or bookformat=='Kindle Edition with Audio/Video' or bookformat=='CD-ROM':
        return 'Audiobook'
    elif bookformat=='Spiral' or bookformat=='Spiral=Bound' or bookformat=='spiral=bound':
        return 'Spiral-Bound'
    elif bookformat=='Boxed Set - Hardcover' or bookformat== 'Bookprint' or bookformat == 'Boxed Collection' or bookformat=='Boxed set':
        return 'Boxed Set'
    elif bookformat=='Library':
        return 'Library Binding'
    elif bookformat=='Comic' or bookformat=='コミック':
        return 'Comics'
    elif bookformat=='Brochura':
        return 'Pamphlet'
    elif bookformat== '78 Card Tarot Deck':
        return 'Deck of 78 Tarot Cards'
    elif bookformat=='Paperback W/Cd' or bookformat=='Paperback with Cd Rom' or bookformat=='Paperback w/ CD' or bookformat=='Paperback with CD Rom':
        return 'Paperback with CD'
    elif bookformat=='Taschenbuch' or bookformat == 'Poche' or bookformat == 'Softcover' or bookformat== 'Soft Cover' or bookformat=='pocket' or bookformat=='soft cover' or bookformat=='softcover':
        return 'Pocket'
    elif bookformat=='Loose Leaf' or bookformat=='Looseleaf' or bookformat=='paper':
        return 'Paper'
    elif bookformat=='3-Ring Binder' or bookformat=='3-ring Binder':
        return 'Ringbound'
    elif bookformat=='Tarot Cards & Book' or bookformat=='Tarot Deck And Booklet' or bookformat== 'Tarot Deck & Booklet' or bookformat=='Tarot Deck and Booklet':
        return 'Tarot Deck & Book'
    elif bookformat=='Paperback with deck of cards':
        return 'Paperback with Deck of Cards'
    elif bookformat=='Paperback and Deluxe Hardcover':
        return 'Paperback And Deluxe Hardcover'
    else:
        return bookformat
    
    
    
    

In [10]:
clean_book['bookformat'] = clean_book['bookformat'].apply(map_bookformat)
clean_book['bookformat'].unique()
#clean_book.to_csv('data/after_encoding/clean_book_data.csv', encoding='utf-8')
clean_book = clean_book.loc[:, ~clean_book.columns.str.contains('^Unnamed')] #delete Unnamed columns pandas
clean_book.to_csv('data/translate_title.csv', encoding='utf-8')

In [4]:
#clean_book = pd.read_csv('data/after_encoding/clean_book_data.csv')
clean_book = pd.read_csv('data/no_extra_space.csv')
len(clean_book['bookformat'].unique())

41

In [13]:
def remove_extra_space(url):
    try:
        title = clean_book.loc[clean_book.link == url, 'title'].values[0]
        clean_book.loc[clean_book.link == url, 'title'] = " ".join(title.split())
        desc = clean_book.loc[clean_book.link == url, 'desc'].values[0]
        clean_book.loc[clean_book.link == url, 'desc'] = " ".join(desc.split())
        author = clean_book.loc[clean_book.link == url, 'author'].values[0]
        clean_book.loc[clean_book.link == url, 'author'] = " ".join(author.split())
        print(title)
    except:
        print(url)

In [14]:
clean_book = pd.read_csv('data/translate_title.csv')
clean_book['link'].apply(remove_extra_space)

Between Two Fires: American Indians in the Civil War
Fashion Sourcebook 1920s
Hungary 56
All-American Anarchist: Joseph A. Labadie and the Labor Movement
The Human Equation: Building Profits by Putting People First
Competitive Advantage Through People: Unleashing the Power of the Work Force
Hawaii: An Uncommon History
Genuine Happiness: Meditation as the Path to Fulfillment
Anthropological Studies of Religion: An Introductory Text
Anarchism And Ecology
Happiness: Lessons from a New Science
Anarchism & Environmental Survival
Dressed to Rule: Royal and Court Costume from Louis XIV to Elizabeth II
On The Medieval Theory Of Signs
After Long Silence
The Principles of Psychology: Volume 1
Islands
Character Strengths and Virtues: A Handbook and Classification
Parnassus on Wheels
The Accidental President of Brazil: A Memoir
The Shining Path: A History of the Millenarian War in Peru
Irrational Exuberance
Jane Austen and the War of Ideas
The Canal House
Toward Rational Exuberance: The Evolution 

0        None
1        None
2        None
3        None
4        None
         ... 
72244    None
72245    None
72246    None
72247    None
72248    None
Name: link, Length: 72249, dtype: object

In [15]:
len(clean_book)

72249

In [ ]:
clean_book = clean_book.loc[:, ~clean_book.columns.str.contains('^Unnamed')] #delete Unnamed columns pandas
clean_book.to_csv('data/no_extra_space.csv',encoding='utf-8')

In [16]:
clean= pd.read_csv('data/translate_title.csv')
len(clean)

72249

In [7]:
genres = dict()

g_og = clean_book.loc[~clean_book['genre'].isnull()]
for _, row in g_og.iterrows():
    genre_list = row['genre'].split(',')
    for g in genre_list:
        if g not in genres:
            genres[g] = 0
        genres[g] += 1

genres

{'History': 11810,
 'Military History': 1494,
 'Civil War': 617,
 'American History': 1756,
 'American Civil War': 347,
 'Nonfiction': 26720,
 'North American Hi...': 1601,
 'Native Americans': 172,
 'Couture': 433,
 'Fashion': 433,
 'Historical': 9202,
 'Art': 4090,
 'Politics': 3297,
 'Labor': 171,
 'Business': 2111,
 'Leadership': 748,
 'Romance': 12478,
 'Historical Romance': 2091,
 'Management': 437,
 'Human Resources': 21,
 'Religion': 5656,
 'Buddhism': 470,
 'Philosophy': 4095,
 'Spirituality': 2537,
 'Psychology': 2791,
 'Anthropology': 837,
 'Academic': 3352,
 'Social Science': 573,
 'Sociology': 1546,
 'Biology': 1072,
 'Ecology': 232,
 'Economics': 1329,
 'Science': 6259,
 'Self Help': 2276,
 'Personal Development': 787,
 'Writing': 2373,
 'Journalism': 290,
 'Cultural': 10627,
 'France': 1278,
 'Autobiography': 3158,
 'Memoir': 3007,
 'World War II': 996,
 'Holocaust': 322,
 'War': 3203,
 'Biography': 5343,
 'Biography Memoir': 1858,
 'Literature': 7172,
 'Jewish': 500,
 '

'North American Hi...': 1601, ---> Problem
 'Scandinavian Literature': 307,
 'Scandinavian Lite...': 97,
 'International Rel...': 33,
 'International Dev...': 14,
 'Science Fiction R...': 27,
 'Democratic Republic Of The ...': 1,
 'Complementary Med...': 7,